### PR TITLE
ESQL: Support the `_source` metadata field

### PR DESCRIPTION
**Pull request by nik9000**
This adds support for load the `_source` field using the syntax:
```
FROM test [METADATA _source]
```

The `_source` field is loaded as a special type - `_source` which no functions support (1). We just render it on the output. Which looks like:
```
$ curl -XDELETE -uelastic:password localhost:9200/test
$ curl -XPOST -HContent-Type:application/json -uelastic:password localhost:9200/test/_doc/1?refresh -d'{
  "words": "words",
  "other stuff": [
    "wow",
    "such",
    "suff"
  ]
}'
$ curl -XPOST -HContent-Type:application/json -uelastic:password localhost:9200/_query?pretty -d'{
  "query": "FROM test [METADATA _source] | KEEP _source | LIMIT 1"
}'

{
  "columns" : [
    {
      "name" : "_source",
      "type" : "_source"
    }
  ],
  "values" : [
    [
      {
        "words" : "words",
        "other stuff" : [
          "wow",
          "such",
          "suff"
        ]
      }
    ]
  ]
}
```

The `_source` is just a json object. We use the same infrastructure to convert it to json as the `_search` response.

This works for both stored `_source` and synthetic `_source`, but it runs row-by-row every time. This *perfect* for stored `_source` but it's less nice to synthetic `_source`. We'd be better of rebuilding synthetic `_source` from blocks but that'd require a lot of new infrastructure. And synthetic `_source` isn't going to be fast anyway.

(1): `IS NULL` and `IS NOT NULL` support `_source` because we get that for free.

### CODE REVIEW

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/nulls/IsNotNullTests.java [hunk 1]
```diff
@@ -10,73 +10,76 @@
...
 import org.hamcrest.Matcher;
 
+import java.util.ArrayList;
 import java.util.List;
 import java.util.function.Supplier;
 
 import static org.hamcrest.Matchers.equalTo;
 
-public class IsNotNullTests extends AbstractScalarFunctionTestCase {
+public class IsNotNullTests extends AbstractFunctionTestCase {
```
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/nulls/IsNotNullTests.java:31]: This was not quite right for _source and we to switch to parameterized tests like this anyway so I just did it now.

**Review by dnhatn — APPROVED** [review]
The PR looks great. Thanks, Nik!

```
  "columns" : [
    {
      "name" : "_source",
      "type" : "_source"
    }
  ],
```

I remembered that you preferred to go with `type: json`. I am curious why you switched to `type: _source`. I am fine with` type: _source`, but I think `type: json` is more interesting, especially because it allows us to develop typeless commands like `jq` in the future.

**Comment by nik9000** [comment]
> I remembered that you preferred to go with `type: json`. I am curious why you switched to `type: _source`. I am fine with` type: _source`, but I think `type: json` is more interesting, especially because it allows us to develop typeless commands like `jq` in the future.

I think I'd always been thinking about `_source` as the type, but I do keep thinking about it as "a json-like thing". I suppose it's not always json - sometimes cbor, yaml, or smile. But it's close enough to json, especially from the language perspective. It's something you could traverse.

I suppose the trouble is that I don't know that you'd really want to traverse it. You get so much from using actual fields and _source traversal is so slow. I get there are cases where you might want some imperative script to traverse it like with a runtime field, but other than that I'm not sure what things you could do with `_source` better than with the fields we already build.

Maybe that's a long way of saying "I'd like to stick with `_source` as the type for now and change it when I know more".

**Comment by dnhatn** [comment]
@nik9000 I think you need to skip the new YAML tests until 8.12.

**Comment by abatsakis** [comment]
> > I remembered that you preferred to go with `type: json`. I am curious why you switched to `type: _source`. I am fine with` type: _source`, but I think `type: json` is more interesting, especially because it allows us to develop typeless commands like `jq` in the future.
> 
> I think I'd always been thinking about `_source` as the type, but I do keep thinking about it as "a json-like thing". I suppose it's not always json - sometimes cbor, yaml, or smile. But it's close enough to json, especially from the language perspective. It's something you could traverse.
> 
> I suppose the trouble is that I don't know that you'd really want to traverse it. You get so much from using actual fields and _source traversal is so slow. I get there are cases where you might want some imperative script to traverse it like with a runtime field, but other than that I'm not sure what things you could do with `_source` better than with the fields we already build.
> 
> Maybe that's a long way of saying "I'd like to stick with `_source` as the type for now and change it when I know more".

How do we treat "object" data types today? Unsupported?

**Comment by nik9000** [comment]
> @nik9000 I think you need to skip the new YAML tests until 8.12.

++

**Comment by nik9000** [comment]
> How do we treat "object" data types today? Unsupported?

They get "flattened" into the fields inside them. You can't explicitly ask for them either.

There are "nested" and `join`fields which are object-like and ESQL also doesn't support them.

I was thinking that we'd grow more functions for picking apparent the _source at some point and we'd have a clearer idea of them as some kind of structure-object-thing.

**Pull request merged**